### PR TITLE
add information for subresource kind determination

### PR DESCRIPTION
**Pull request by deads2k**
xref https://github.com/kubernetes/kubernetes/issues/38810 https://github.com/kubernetes/kubernetes/issues/38756

Polymorphic subresources usually have different groupVersions for their discovery kinds than their "native" groupVersions.  Even though the APIResourceList shows the kind properly, it does not reflect the group or version of that kind, which makes it impossible to unambiguously determine if the subresource matches you and it is impossible to determine how to serialize your data.  See HPA controller.

This adds an optional Group and Version to the discovery doc, which can be used to communicate the "native" groupversion of an endpoint.  Doing this does not preclude fancier contenttype negotiation in the future and doesn't prevent future expansion from indicating equivalent types, but it does make it possible to solve the problem we have today or polymorphic categorization.

@kubernetes/sig-api-machinery-misc @smarterclayton 
@cheftako since @lavalamp is out.

```release-note
Adds optional group and version information to the discovery interface, so that if an endpoint uses non-default values, the proper value of "kind" can be determined. Scale is a common example.
```

### CODE REVIEW

**Comment by deads2k** [comment]
This pull just shows the new fields without wiring or generation, so we can first sort out how to express it.

@DirectXMan12 you want this too I think.

**Comment by DirectXMan12** [comment]
woot!

This will eventually unblock me on a generic scale client.

**Comment by smarterclayton** [comment]
Seems very reasonable to me, do we need to qualify group / version with a prefix in order to distinguish them better?

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go [hunk 1]
```diff
@@ -750,6 +750,12 @@ type APIResource struct {
 	SingularName string `json:"singularName" protobuf:"bytes,6,opt,name=singularName"`
 	// namespaced indicates if a resource is namespaced or not.
 	Namespaced bool `json:"namespaced" protobuf:"varint,2,opt,name=namespaced"`
+	// group is the preferred group of the resource.  Empty implies the group of the containing resource list.
+	// For subresources, this may have a different value, for example: Scale".
+	Group string `json:"group,omitempty" protobuf:"bytes,8,opt,name=group"`
+	// version is the preferred version of the resource.  Empty implies the version of the containing resource list
+	// For subresources, this may have a different value, for example: Scale".
```
Comment (cheftako) [inline staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go]: quibble but can we fix the example to be a version?

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go [hunk 2]
```diff
@@ -750,6 +750,12 @@ type APIResource struct {
 	SingularName string `json:"singularName" protobuf:"bytes,6,opt,name=singularName"`
 	// namespaced indicates if a resource is namespaced or not.
 	Namespaced bool `json:"namespaced" protobuf:"varint,2,opt,name=namespaced"`
+	// group is the preferred group of the resource.  Empty implies the group of the containing resource list.
```
Comment (cheftako) [inline staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go]: Don't we expect there times when there are multiple versions of the APIResource? If we want to switch groups then there could even be multiple groups? Would preferred be the most stable? I assume that the preferred group must match the preferred version?

Comment (deads2k) [inline staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go]: > I assume that the preferred group must match the preferred version?

I think this group and version should always match and be the preferred one (the one that most people will use.  Our internal code already handles this case today.



> Don't we expect there times when there are multiple versions of the APIResource?

Eventually, yes.  Right now our code doesn't handle the condition and explicitly forbids it.  I think that we should reflect what we have now (and realistically for the next few releases).



> If we want to switch groups then there could even be multiple groups? Would preferred be the most stable?

I think we would handle this as we've done for our preferred versions.  We would add an array of possible GroupVersions and say that they are in the preferred ordering, leaving the choice to the individual api server.

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go (hunk 1 above)
Comment (deads2k) [inline staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go]: > quibble but can we fix the example to be a version?

Yes.  I'll update during my rebase and generation cleanup today.

**Comment by deads2k** [comment]
Updated for comments and generated code.

**Comment by deads2k** [comment]
@kubernetes/api-approvers

**Comment by liggitt** [comment]
> @smarterclayton 
> Seems very reasonable to me, do we need to qualify group / version with a prefix in order to distinguish them better?

Seems reasonable to me as well. I don't think a prefix is needed here... other discovery types use unprefixed `groups`, `versions`, and `version` field names

**Comment by smarterclayton** [comment]
Any other comments?  I think i'm ok with this as is.

**Comment by smarterclayton** [comment]
/approve

based on general agreement from participants

**Pull request merged**